src/wallet/spend.cpp

```diff
@@ -1392,4 +1392,293 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
+{
+    if (inputs.empty()) {
+        return util::Error{_("Inputs must not be empty")};
+    }
+
+    CCoinControl coin_control = SetupDeniabilizationCoinControl(confirm_target);
+    for (const auto& input : inputs) {
+        coin_control.Select(input);
+    }
+    Assert(coin_control.HasSelected());
+    CFeeRate deniabilization_fee_rate = CalculateDeniabilizationFeeRate(wallet, confirm_target);
+    coin_control.m_feerate = deniabilization_fee_rate;
+
+    LOCK(wallet.cs_wallet);
+
+    FastRandomContext rng_fast;
+    CoinSelectionParams coin_selection_params{rng_fast};
+    auto res_fetch_inputs = FetchSelectedInputs(wallet, coin_control, coin_selection_params);
+    if (!res_fetch_inputs) {
+        return util::Error{util::ErrorString(res_fetch_inputs)};
+    }
+    PreSelectedInputs preset_inputs = *res_fetch_inputs;
+    CAmount total_amount = preset_inputs.total_amount;
+
+    // validate that all UTXOs share the same address
+    std::optional<CScript> op_shared_script;
+    for (const auto& coin : preset_inputs.coins) {
+        if (!op_shared_script) {
+            op_shared_script = coin->txout.scriptPubKey;
+        }
+        if (!op_shared_script || !(*op_shared_script == coin->txout.scriptPubKey)) {
+            return util::Error{_("Input addresses must all match.")};
+        }
+    }
+    Assert(op_shared_script);
+    CScript shared_script = *op_shared_script;
+
+    CFeeRate discard_feerate = GetDiscardRate(wallet);
+    CAmount dust_threshold = GetDustThreshold(CTxOut(total_amount, shared_script), discard_feerate);
+
+    // deniabilize the UTXOs by splitting the value randomly
+    // find a split that leaves enough amount post split to finish the deniabilization process in each new UTXO
+    CAmount min_post_split_amount = CalculateDeniabilizationFeeEstimate(shared_script, total_amount / NUM_DENIABILIZATION_OUTPUTS, 1, deniabilization_cycles + 1, deniabilization_fee_rate) + dust_threshold;
+    CAmount estimated_tx_fee = CalculateDeniabilizationTxFee(shared_script, total_amount, preset_inputs.coins.size(), deniabilization_fee_rate);
+
+    CAmount total_random_range = total_amount - min_post_split_amount * NUM_DENIABILIZATION_OUTPUTS - estimated_tx_fee;
+    if (total_random_range < 0) {
+        insufficient_amount = true;
+        return util::Error{strprintf(_("Insufficient amount (%d) for a deniabilization transaction, min amount (%d), tx fee (%d)."), total_amount, min_post_split_amount, estimated_tx_fee)};
+    }
+
+    CTxDestination shared_destination = CNoDestination();
+    ExtractDestination(shared_script, shared_destination);
+    std::optional<OutputType> op_output_type = OutputTypeFromDestination(shared_destination);
+    if (!op_output_type) {
+        op_output_type = wallet.m_default_change_type;
+    }
+    OutputType output_type = op_output_type ? *op_output_type : wallet.m_default_address_type;
```

Comment:
@achow101 m
Is this a correct way to get an OutputType from a CScript? I need the OutputType for the ReserveDestination constructor.
For coinbase key wallet scripts, the destination gets classified as PubKeyDestination, thus OutputTypeFromDestination will return nullopt, so we'll end up using the wallet.m_default_address_type.
This appears to work, but I'm not sure if it's kosher ...

Reply:
That looks fine.